src/app/InteractionModelEngine.cpp

```diff
@@ -289,6 +293,42 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
     return err;
 }
 
+CHIP_ERROR InteractionModelEngine::OnUnsolicitedReportData(Messaging::ExchangeContext * apExchangeContext,
+                                                           const PacketHeader & aPacketHeader, const PayloadHeader & aPayloadHeader,
+                                                           System::PacketBufferHandle && aPayload)
+{
+    System::PacketBufferTLVReader reader;
+    reader.Init(aPayload.Retain());
+    ReturnLogErrorOnFailure(reader.Next());
+
+    ReportData::Parser report;
+    ReturnLogErrorOnFailure(report.Init(reader));
+
+    uint64_t subscriptionId = 0;
+    ReturnLogErrorOnFailure(report.GetSubscriptionId(&subscriptionId));
+
+    for (auto & readClient : mReadClients)
+    {
+        if (!readClient.IsSubscriptionTypeIdle())
+        {
+            continue;
+        }
+        if (!readClient.IsMatchingClient(subscriptionId))
+        {
+            continue;
+        }
+
+        readClient.SetExchangeContext(apExchangeContext);
+        CHIP_ERROR err = readClient.ProcessReportData(std::move(aPayload));
+        if (err != CHIP_NO_ERROR)
+        {
+            readClient.Shutdown();
```

Comment:
Is there a situation where we want 'error in processreportdata should not shutdown'?

I would find a `return readClient.ProcessReportData(std::move(aPayload));` more obvious to read than this.

Reply:
move shutdown in ProcessReportData , the lifetime is controlled inside readclient, done, thanks